selfservice/strategy/code/strategy.go

```diff
@@ -82,14 +119,348 @@ func NewStrategy(deps strategyDependencies) *Strategy {
+		*login.Flow | *registration.Flow | *recovery.Flow | *verification.Flow
+		flow.Flow
+	}
+	FlowPayload interface {
+		*updateLoginFlowWithCodeMethod | *updateRegistrationFlowWithCodeMethod | *updateRecoveryFlowWithCodeMethod | *updateVerificationFlowWithCodeMethod
+	}
+	CreateCodeState[T FlowType, P FlowPayload]           func(context.Context, T, *Strategy, P) error
+	ValidateCodeState[T FlowType, P FlowPayload]         func(context.Context, T, *Strategy, P) error
+	AlreadyValidatedCodeState[T FlowType, P FlowPayload] func(context.Context, T, *Strategy, P) error
+	CodeStateManager[T FlowType, P FlowPayload]          struct {
+		f                         T
+		payload                   P
+		strategy                  *Strategy
+		createCodeState           CreateCodeState[T, P]
+		verifyCodeState           ValidateCodeState[T, P]
+		alreadyValidatedCodeState AlreadyValidatedCodeState[T, P]
+	}
+)
+
+func NewCodeStateManager[T FlowType, P FlowPayload](f T, s *Strategy, payload P) *CodeStateManager[T, P] {
+	return &CodeStateManager[T, P]{
+		f:        f,
+		strategy: s,
+		payload:  payload,
+	}
+}
+
+func (c *CodeStateManager[T, P]) SetCreateCodeHandler(fn CreateCodeState[T, P]) {
+	c.createCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeVerifyHandler(fn ValidateCodeState[T, P]) {
+	c.verifyCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeDoneHandler(fn AlreadyValidatedCodeState[T, P]) {
+	c.alreadyValidatedCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) validatePayload(ctx context.Context) error {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		if len(v.Identifier) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/identifier", "identifier"))
+		}
+	case *updateRegistrationFlowWithCodeMethod:
+		if len(v.Traits) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/traits", "traits"))
+		}
+	case *updateRecoveryFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	case *updateVerificationFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	default:
+		return errors.WithStack(herodot.ErrBadRequest.WithReason("received unexpected flow payload type"))
+	}
+	return nil
+}
+
+func (c *CodeStateManager[T, P]) getResend() string {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		return v.Resend
+	case *updateRegistrationFlowWithCodeMethod:
+		return v.Resend
+	}
+	return ""
+}
+
+func (c *CodeStateManager[T, P]) getCode() string {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		return v.Code
+	case *updateRegistrationFlowWithCodeMethod:
+		return v.Code
+	case *updateRecoveryFlowWithCodeMethod:
+		return v.Code
+	case *updateVerificationFlowWithCodeMethod:
+		return v.Code
+	}
+	return ""
+}
+
+func (c *CodeStateManager[T, P]) Run(ctx context.Context) error {
```

Comment:
I think this is reasonably valid, one problem with this though is that we don't know (on the outside) whether the correct callback was indeed called. Did callback "verifyCodeState" really get executed? We can't tell! We only know that we received no error.

Apart from a couple of similar checks (is state empty? is code empty?) we seem to be building more logic for this abstraction.

But is the abstraction even needed? I'm not sure if it really saves us a lot of code (which seems to be the purpose) and it also doesn't have dedicated tests that prove its correctness, in which case we might just rely on the test paths in strategy_login_test etc.

Reply:
I think the purpose is two fold.
- readability inside the actual implementation. In the recovery flow for example, it's tough to know which part is for which state of the flow. Are we handling a first submission (flow state `choose_method`) or are we handling a second submission? Where is it exiting the flow on an error is also not too apparent and where was it successful? 

- Verification, Recovery, Login, Registration and Settings will all have the same states:
 1. choose_method (fist submit with traits / email)
 2. Error? retunr back to choose_method
 3. email_sent (second submit with code)
 4. Error? return back to email_sent 
 5. challenge_passed (last submit)
 6. Resubmit passed challenge? Should error immediately. 
 
 
This code was more to manage these states across all of the flows. And since each flow has its own logic, we do the callback code. I guess we could make this one large method with a switch case with each state and flow, but this would be hard to read as well.